xds/src/main/java/io/grpc/xds/ClusterManagerLoadBalancer.java

```diff
@@ -0,0 +1,263 @@
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.ClusterManagerLoadBalancerProvider.ClusterManagerConfig;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * The top-level load balancing policy.
+ */
+class ClusterManagerLoadBalancer extends LoadBalancer {
+
+  @VisibleForTesting
+  static final int DELAYED_ACTION_DELETION_TIME_MINUTES = 15;
+  @VisibleForTesting
+  static final CallOptions.Key<String> ROUTING_CLUSTER_NAME_KEY =
+      CallOptions.Key.create("io.grpc.xds.ROUTING_CLUSTER_NAME_KEY");
+
+  private final Map<String, ChildLbState> childLbStates = new HashMap<>();
+  private final Helper helper;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService timeService;
+  private final XdsLogger logger;
+
+  ClusterManagerLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+    this.syncContext = checkNotNull(helper.getSynchronizationContext(), "syncContext");
+    this.timeService = checkNotNull(helper.getScheduledExecutorService(), "timeService");
+    logger = XdsLogger.withLogId(InternalLogId.allocate("cds-lb", helper.getAuthority()));
+    logger.log(XdsLogLevel.INFO, "Created");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+    ClusterManagerConfig config = (ClusterManagerConfig)
+        resolvedAddresses.getLoadBalancingPolicyConfig();
+    Map<String, PolicySelection> newChildPolicies = config.childPolicies;
+    logger.log(
+        XdsLogLevel.INFO,
+        "Received cluster_manager lb config: child names={0}", newChildPolicies.keySet());
+    for (Map.Entry<String, PolicySelection> entry : newChildPolicies.entrySet()) {
+      final String name = entry.getKey();
+      LoadBalancerProvider childPolicyProvider = entry.getValue().getProvider();
+      Object childConfig = entry.getValue().getConfig();
+      if (!childLbStates.containsKey(name)) {
+        childLbStates.put(name, new ChildLbState(name, childPolicyProvider));
+      } else {
+        childLbStates.get(name).reactivate(childPolicyProvider);
+      }
+      final ResolvedAddresses childAddresses =
+          resolvedAddresses.toBuilder().setLoadBalancingPolicyConfig(childConfig).build();
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          childLbStates.get(name).lb.handleResolvedAddresses(childAddresses);
+        }
+      });
+    }
+    for (String name : childLbStates.keySet()) {
+      if (!newChildPolicies.containsKey(name)) {
+        childLbStates.get(name).deactivate();
+      }
+    }
```

Comment:
From our last discussion, we seemed to agree on updating the picker immediately if the list of clusters is changed, otherwise new RPCs may be routed to a new cluster but the picker does not include that cluster and will end up with failure.

Reply:
Oh, you are right, but that would be something slightly different. I think we still want the delayed deletion mechanism for the balancer subtree. However, when receiving an updated config, we need to immediately propagate an updated picker that will not choose removed child balancers. Currently, the new picker will only be propagated up when one of subchannels state changes. Note, the new picker will not choose removed/deactivated child balancers (see line 138), but the problem is we still need an immediate picker update. I will fix this.

Reply:
I updated the implementation, as well as refactored the tests. PTAL.